relay-server/src/actors/envelopes.rs

```diff
@@ -1972,61 +2119,122 @@ impl Handler<HandleEnvelope> for EnvelopeManager {
         let event_id = envelope.event_id();
         let remote_addr = envelope.meta().client_addr();
 
-        let scoping = Rc::new(RefCell::new(envelope.meta().get_partial_scoping()));
-        let is_received = Rc::new(AtomicBool::from(false));
-        let envelope_summary = Rc::new(RefCell::new(EnvelopeSummary::compute(&envelope)));
+        let envelope_context = Rc::new(RefCell::new(EnvelopeContext::new(
+            EnvelopeSummary::compute(&envelope),
+            relay_common::instant_to_date_time(start_time),
+            event_id,
+            remote_addr,
+            envelope.meta().get_partial_scoping(),
+        )));
 
         let future = project_cache
             .send(CheckEnvelope::fetched(project_key, envelope))
             .map_err(ProcessingError::ScheduleFailed)
             .and_then(|result| result.map_err(ProcessingError::ProjectFailed))
-            .and_then(clone!(scoping, envelope_summary, |response| {
+            .map_err(clone!(envelope_context, outcome_producer, |err| {
+                if let Some(outcome) = err.to_outcome() {
+                    envelope_context
+                        .borrow()
+                        .send_outcomes(outcome, outcome_producer);
+                }
+                err
+            }))
+            .and_then(clone!(envelope_context, outcome_producer, |response| {
                 // Use the project id from the loaded project state to account for redirects.
                 let project_id = response.scoping.project_id.value();
                 metric!(set(RelaySets::UniqueProjects) = project_id as i64);
 
-                scoping.replace(response.scoping);
+                envelope_context.borrow_mut().set_scoping(response.scoping);
+
+                let checked = match response.result {
+                    Err(err) => {
+                        let err = ProcessingError::Rejected(err);
+                        if let Some(outcome) = err.to_outcome() {
+                            envelope_context
+                                .borrow()
+                                .send_outcomes(outcome, outcome_producer);
```

Comment:
Please move this and the following check for `None` into handling of `CheckEnvelope`. Same with sampling.